src/core.rs

```diff
@@ -502,30 +536,188 @@ pub enum WrapAlgorithm {
+    W: Fn(usize) -> usize,
+{
+    WrapFirstFit {
+        inner: WrapFirstFitInner {
+            fragments: fragments.into_iter(),
+            line_widths,
+            line_number: 0,
+            remaining_width: None,
+        }
+        .peekable(),
+        terminating_eol: false,
+    }
+}
+
+/// Iterator for [`wrap_first_fit`].
+#[must_use = "iterators are lazy and do nothing unless consumed"]
+pub struct WrapFirstFit<F, W>
+where
+    F: Iterator,
+    <F as Iterator>::Item: Fragment,
+    W: Fn(usize) -> usize,
+{
+    inner: std::iter::Peekable<WrapFirstFitInner<F, W>>,
+    terminating_eol: bool,
+}
+
+impl<F, W> WrapFirstFit<F, W>
+where
+    F: Iterator,
+    <F as Iterator>::Item: Fragment,
+    W: Fn(usize) -> usize,
+{
+    /// Add a terminating EOL (i.e, have the last fragment return
+    /// `true` in its tuple).
+    pub fn terminate_eol(self) -> Self {
+        Self {
+            terminating_eol: true,
+            ..self
+        }
+    }
+}
+
+impl<F, W> Iterator for WrapFirstFit<F, W>
+where
+    F: Iterator,
+    <F as Iterator>::Item: Fragment,
+    W: Fn(usize) -> usize,
+{
+    type Item = (<F as Iterator>::Item, bool);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let (_, fragment) = self.inner.next()?;
+        let eol = self
+            .inner
+            .peek()
+            .map(|&(newline, _)| newline)
+            .unwrap_or(self.terminating_eol);
+        Some((fragment, eol))
+    }
+}
+impl<F, W> ExactSizeIterator for WrapFirstFit<F, W>
+where
+    F: ExactSizeIterator,
+    <F as Iterator>::Item: Fragment,
+    W: Fn(usize) -> usize,
+{
+    fn len(&self) -> usize {
+        self.inner.len()
+    }
+}
+
+impl<F, W> FusedIterator for WrapFirstFit<F, W>
+where
+    F: FusedIterator,
+    <F as Iterator>::Item: Fragment,
+    W: Fn(usize) -> usize,
+{
+}
+
+// Manual impl so that `<F as Iterator>::Item: Debug`.
+impl<F, W> Debug for WrapFirstFit<F, W>
+where
+    F: Iterator + Debug,
+    <F as Iterator>::Item: Fragment + Debug,
+    W: Fn(usize) -> usize + Debug,
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.debug_struct("WrapFirstFit")
+            .field("inner", &self.inner)
+            .field("terminating_eol", &self.terminating_eol)
+            .finish()
+    }
+}
 
-    for (idx, fragment) in fragments.iter().enumerate() {
-        let line_width = line_widths(lines.len());
-        if width + fragment.width() + fragment.penalty_width() > line_width && idx > start {
-            lines.push(&fragments[start..idx]);
-            start = idx;
-            width = 0;
+// Manual impl so that `<F as Iterator>::Item: Clone`.
+impl<F, W> Clone for WrapFirstFit<F, W>
+where
+    F: Iterator + Clone,
+    <F as Iterator>::Item: Fragment + Clone,
+    W: Fn(usize) -> usize + Clone,
+{
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+            terminating_eol: self.terminating_eol,
+        }
+    }
+}
```

Comment:
If this isn't needed, then please drop it. I'm not sure what the semantics would be of cloning an `Iterator` would be? Does advancing one iterator also advance the other?

Reply:
Implementing `Clone` for iterators is very common - the standard library does it for nearly every one of its iterator types and itertools does too. I don't see the benefit in _not_ having this - having it could be useful and it's unlikely someone will go out of their way to implement this at any point, so why not make your library slightly more useful?

Reply:
I see that you're right — the `Clone` and `Debug` implementations are everywhere in the standard library... I hadn't noticed them until now.

The benefit of not providing the implementation is the same as usual: less code to maintain, less constraints on future developments. Less API surface to maintain also results in less noise in the documentation. I consider things like [`option::Iter`](https://doc.rust-lang.org/std/option/struct.Iter.html), [`result::Iter`](https://doc.rust-lang.org/std/result/struct.Iter.html), etc noise. If the stdlib had been written today, I fully expect that it would have hidden all these structs behnd `-> impl Iterator + Debug + ...`.

Reply:
I guess I just dislike code :smile: The less, the better, in my opinion. I feel better about the code if I know what problem it solves, and I've always tried to avoid solving potential problems that haven't actually materialized themselves yet.

That being said, I guess these implementations aren't so important and I guess they can stay. It would just be nice if you had an example program that needs this code somehow.